packages/modules/vehicles/leaf/soc.py

```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env python3
+from typing import List
+
+import logging
+import time
+
+from helpermodules.cli import run_using_positional_cli_args
+from modules.common import store
+from modules.common.abstract_device import DeviceDescriptor
+from modules.common.abstract_vehicle import VehicleUpdateData
+from modules.common.component_state import CarState
+from modules.common.configurable_vehicle import ConfigurableVehicle
+from modules.vehicles.leaf.config import LeafSoc, LeafConfiguration
+
+import pycarwings2
+
+log = logging.getLogger(__name__)
+
+
+def fetch_soc(username, password, chargepoint) -> CarState:
+
+    region = "NE"
+
+    def getNissanSession():     # open Https session with Nissan server
+        log.debug("LP%s: login = %s, region = %s" % (chargepoint, username, region))
+        s = pycarwings2.Session(username, password, region)
+        leaf = s.get_leaf()
+        time.sleep(1)           # give Nissan server some time
+        return leaf
+
+    def readSoc(leaf):          # get SoC from Nissan server
+        leaf_info = leaf.get_latest_battery_status()
+        bat_percent = int(leaf_info.battery_percent)
+        log.debug("LP%s: Battery status %s" % (chargepoint, bat_percent))
+        return bat_percent
+
+    def requestSoc(leaf):       # request Nissan server to request last SoC from car
+        log.debug("LP%s: Request SoC Update" % (chargepoint))
+        key = leaf.request_update()
+        status = leaf.get_status_from_update(key)
+        sleepsecs = 20
+        while status is None:
+            log.debug("Waiting {0} seconds".format(sleepsecs))
+            time.sleep(sleepsecs)
+            status = leaf.get_status_from_update(key)
```

Comment:
```suggestion
        for i in range(0,3):
            log.debug("Waiting {0} seconds".format(sleepsecs))
            time.sleep(sleepsecs)
            status = leaf.get_status_from_update(key)
            if status is not None:
                break
```
Wenn der Server nicht erreichbar ist, führt das while zu einer Endlosschleife.

Reply:
Diese Zeilen in der Funktion requestSoC würde ich nur ungern anfassen. Die Funktionen getNissanSession, readSoC und requestSoC wurden 1:1 aus dem Modul soc.py von V1.9 übernommen und in fetch_soc eingebunden. Bislang lief die Funktion requestSoC so auch in V1.9.  
Die Verbindung zum Server besteht zu dem Zeitpunkt bereits. Innerhalb der von requestSoC aufgerufen Funktionen get_status_from_update etc. wird auf Status >400 im Reply geachtet und auch Excemptions ausgelöst. Notfalls wird in _request_with_retry sogar ein neuer connect ausgelöst. Früher oder später sollte bei diesen Abfragen und Versuchen ein Status zurückkommen. 
Zwei bis drei Zyklen a' 20 Sekunden ist eher normal als selten beim Leaf. Der Server muss den Leaf über Internet und Mobilfunknetz erreichen. Da kann es einige Fehlversuche geben, bis das klappt. Bei Misserfolg gibt der Server sicherlich auch eine dazu passende Antwort zurück. 
Falls wir die Schleife doch endlich machen, dann auf jeden Fall mit mehr als drei Zyklen, z.B. 10. 
Bitte um eine Rückmeldung zu meiner Antwort bevor ich den Code anfasse.